src/_P090_MitsubishiHeatPump .ino

```diff
@@ -0,0 +1,698 @@
+  for (int i = 0; i < len; i++) {
+    if (byteMap[i] == byteValue) {
+      return valuesMap[i];
+    }
+  }
+  return valuesMap[0];
+}
+
+int lookupByteMapIndex(const int valuesMap[], int len, int lookupValue) {
+  for (int i = 0; i < len; i++) {
+    if (valuesMap[i] == lookupValue) {
+      return i;
+    }
+  }
+  return -1;
+}
+
+int lookupByteMapIndex(const char* valuesMap[], int len, const char* lookupValue) {
+  for (int i = 0; i < len; i++) {
+    if (strcasecmp(valuesMap[i], lookupValue) == 0) {
+      return i;
+    }
+  }
+  return -1;
+}
+
+bool lookupValue(const char* valuesMap[], int len, const char* lookupValue, const char*& destination) {
+  int index = lookupByteMapIndex(valuesMap, len, lookupValue);
+  if (index == -1) {
+    return false;
+  }
+  destination = valuesMap[index];
+  return true;
+}
+
+byte checkSum(byte bytes[], int len) {
+  byte sum = 0;
+  for (int i = 0; i < len; i++) {
+    sum += bytes[i];
+  }
+  return (0xfc - sum) & 0xff;
+}
+
+static bool operator!=(const heatpumpSettings& lhs, const heatpumpSettings& rhs) {
+  return lhs.power != rhs.power ||
+         lhs.mode != rhs.mode ||
+         lhs.temperature != rhs.temperature ||
+         lhs.fan != rhs.fan ||
+         lhs.vane != rhs.vane ||
+         lhs.wideVane != rhs.wideVane ||
+         lhs.iSee != rhs.iSee;
+}
+
+P090_data_struct::P090_data_struct(const int16_t serialRx, const int16_t serialTx) :
+  _serial(serialRx, serialTx),
+  _currentSettings({}),
+  _currentStatus({}),
+  _isConnected(false),
+  _bitrate(2400),
+  _waitForRead(false),
+  _lastSend(0),
+  _lastRecv(millis() - (PACKET_SENT_INTERVAL_MS * 10)),
+  _tempMode(false),
+  _wideVaneAdj(false),
+  _infoMode(0),
+  _isUpdatingSettings(false) {
+
+}
+
+bool P090_data_struct::write(const String& command, const String& value) {
+  if (_isUpdatingSettings) {
+    return false;
+  }
+
+  _isUpdatingSettings = true;
+
+  bool success = false;
+  heatpumpSettings settings = _currentSettings;
+
+  if (command == F("temperature")) {
+    success = string2float(value, settings.temperature);
+  } else if (command == F("power")) {
+    success = lookupValue(POWER_MAP, 2, value.c_str(), settings.power);
+  } else if (command == F("mode")) {
+    success = lookupValue(MODE_MAP, 5, value.c_str(), settings.mode);
+  } else if (command == F("fan")) {
+    success = lookupValue(FAN_MAP, 6, value.c_str(), settings.fan);
+  } else if (command == F("vane")) {
+    success = lookupValue(VANE_MAP, 7, value.c_str(), settings.vane);
+  } else if (command == F("widevane")) {
+    success = lookupValue(WIDEVANE_MAP, 7, value.c_str(), settings.wideVane);
+  }
+
+  if (success && update(settings)) {
+    _currentSettings = settings;
+  } else {
+    success = false;
+  }
+
+  _isUpdatingSettings = false;
+
+  return success;
+}
+
+void P090_data_struct::connect(bool retry) {
+  _isConnected = false;
+  _serial.begin(_bitrate, SERIAL_8E1);
+
+  writePacket(CONNECT, CONNECT_LEN);
+  while(!canRead()) { delay(10); }
```

Comment:
Is this really needed?
This is blocking code and that's something I'm working really hard on to get rid of blocking code.
Also you're not using a timeout, so if the connect function is entered without the device responding, the ESP will hang.
You should use a state machine and check in the PLUGIN_TEN_PER_SECOND call if you can go to the next state.